### ISSUE TITLE
Support XDG FileChooser portal

### ISSUE THREAD

**Issue by TheEvilSkeleton** [tenacityteam/tenacity-legacy#436]
- [x] I have read the specified guidelines for issues

**Is your feature request related to a problem? Please describe.**

One of the problems with the Linux desktop is file pickers. Whenever we use a(n old) GTK app on Plasma, we get the GTK file picker, and vice-versa.

Tenacity uses the GTK file picker, so matter what DE you use, it will use the GTK file picker.

**Describe the solution you'd like**

I'd like to suggest supporting [XDG FileChooser portal](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.FileChooser). This will make the app open the file picker of the DE you are using, which will make it a lot more consistent to the desktop and more familiar.

It's also really beneficial for Flatpak users, because this will let us remove `filesystem=host` permission (read write permissions to your whole system), which will make the Tenacity Flatpak infinitely more secure than what it is now. `filesystem=host` is as secure as installing a native package, which means it's inherently insecure; letting it access your shell configuration files, GPG and SSH keys, passwords, etc. is extremely risky and generally shouldn't be advisable generally, whether it's FOSS or not, because security vulnerabilities exist and can be exploited. Sadly, Tenacity is an audio editor, so it's crucial to easily access your files as well.

~~I'd also suggest supporting other portals whenever possible. This will improve the upcoming flatpak by a lot because we'll be able to remove `filesystem=host` permission and fully harden the application.~~

**Comment by Be-ing** [comment]
> I'd also suggest supporting other portals whenever possible. This will improve the upcoming flatpak by a lot because we'll be able to remove filesystem=host permission and fully harden the application.

Please open separate issues for each of those.

**Comment by n0toose** [comment]
- "Qt and KDE libraries will transparently use portals for some functionality when they detect that they are being used inside a Flatpak sandbox."

  https://docs.flatpak.org/en/latest/portals-qt.html

- Audacity/Tenacity ships with a [default/custom FileDialog.cpp widget](https://github.com/tenacityteam/tenacity/blob/master/src/widgets/FileDialog/FileDialog.cpp).
- Audacity/Tenacity ships custom File Dialogs for [Windows, macOS and GTK](https://github.com/tenacityteam/tenacity/tree/d28166652c33d9f81ac63ef3767852985049462b/src/widgets/FileDialog). They also handle the relevant permissions.

**Comment by n0toose** [comment]
Keep in mind that the dialogs use private APIs in OS X. We will have to do away with those if we ever plan to release Tenacity on the App Store, as Apple specifically prohibits this for apps published in their store. I think this is why [f.lux still needs to be manually downloaded](https://justgetflux.com/news/pages/macquickstart/) on that platform.

**Comment by n0toose** [comment]
I think the FileDialog thing is throwing me off. Let's start from the beginning.

- The "Open..." button runs the function OnOpen.

https://github.com/tenacityteam/tenacity/blob/a0ea8b63ad472b9b2b21107b662a39767be03ae6/src/menus/FileMenus.cpp#L567-L568

- OnOpen uses `ProjectManager::OpenFiles(project);`, which actually uses another function in `src/ProjectFileManager.cpp` called `ShowOpenDialog`, which actually uses a wxWidgets-adjacent class called [FileDialogWrapper](https://github.com/tenacityteam/tenacity/blob/d28166652c33d9f81ac63ef3767852985049462b/src/widgets/wxPanelWrapper.h#L163)... I am not completely sure what the [FileDialog.cpp](https://github.com/tenacityteam/tenacity/blob/master/src/widgets/FileDialog/FileDialog.cpp) that was originally shared is all about, it's very likely relevant, and I'm not sure how much of a role inherited classes from the GTK/Qt APIs play here.

https://github.com/tenacityteam/tenacity/blob/a0ea8b63ad472b9b2b21107b662a39767be03ae6/src/menus/FileMenus.cpp#L180-L184
https://github.com/tenacityteam/tenacity/blob/44968d3ac316e4a4fd465cb02976fc51699b44fc/src/ProjectManager.cpp#L856-L859

**Comment by Be-ing** [comment]
Shouldn't this be fixed upstream in wxWidgets? Why are we reinventing the wheel here...

**Comment by n0toose** [comment]
I am not sure, because I am not completely sure whether we fully rely on their own stuff or whether Audacity has come up with its own dialogs. Problem is, if we actually go to wxWidgets and push a patch ourselves, it'll take ages until that lands in a version that we can actually use, but I guess that's the easiest way for all of us, including people working on other wxWidgets-related projects, really.

**Comment by n0toose** [comment]
Either way, we have to figure out if the "onus is on us" first.

**Comment by Be-ing** [comment]
We can use whatever version of wxWidgets we want in the Flatpak, including a fork.

**Comment by n0toose** [comment]
Again, we have to figure out whether wxWidgets needs to be modified at all, or if we have some custom dialog stuff to deal with by ourselves.

**Comment by Be-ing** [comment]
We should get rid of any custom dialog stuff and work upstream in wxWidgets.

**Comment by TheEvilSkeleton** [comment]
I had a chat with some people that are closely tied with Flatpak and XDG, and they suggested using libportal https://github.com/flatpak/libportal.

**Comment by leio** [comment]
Or we just switch the already copied wxFileDialog gtk implementation in src/widgets/FileDialog/FileDialogPrivate.cpp to use GtkFileChooserNative, if we don't need the full dialogs features. Of course it would also make sense to sync with upstream, clean up ancient hacks with (these days) dubious reasoning and try to upstream the end result changes to wxWidgets as well.

Maybe I can take a look during some vacation. You can find my name from the header of that file already anyways, afterall :D

**Comment by leio** [comment]
One thing I don't know though - will GtkFileChooserNative usage give us native Qt dialog under Plasma also when not running via flatpak, or only if there's some special desktop env setup magic sauce to tell gtk to do that, or not at all and we'd need libportal in that case anyways?

**Comment by TheEvilSkeleton** [comment]
> Gtk.FileChooserNative is an abstraction of a dialog box suitable for use with “File/Open” or “File/Save as” commands. By default, this just uses a Gtk.FileChooserDialog to implement the actual dialog. However, on certain platforms, such as Windows and macOS, the native platform file chooser is used instead. When the application is running in a sandboxed environment without direct filesystem access (such as Flatpak), Gtk.FileChooserNative may call the proper APIs (portals) to let the user choose a file and make it available to the application.

https://lazka.github.io/pgi-docs/Gtk-3.0/classes/FileChooserNative.html#Gtk.FileChooserNative

So this is the one that uses the DE's native file picker if you use the Flatpak package?

**Comment by leio** [comment]
yes, https://docs.gtk.org/gtk3/class.FileChooserNative.html or https://developer-old.gnome.org/gtk3/stable/gtk3-GtkFileChooserNative.html (it doesn't link to the GtkFileChooser interface there that has most of the methods, just like GtkFileChooserDialog)

**Comment by Semisol** [comment]
> if we ever plan to release Tenacity on the App Store

Weren't there some license requirements...?

**Comment by TheEvilSkeleton** [comment]
@leio I forgot to mention that we'll probably have to use libportal temporarily because, based on my friends, they said:

> yeah looks like wxWidgets just uses GtkFileChooserDialog not GtkFileChooserNative
> so you'd have to use libportal

If we manage to send the relevant patches to upstream, we might have FileChooser portal without needing to use libportal, from what I understood.

> > if we ever plan to release Tenacity on the App Store
> 
> Weren't there some license requirements...?

Based on https://developer.apple.com/forums/thread/18922, we cannot release it in the App Store, which sadly makes Audacity the only viable option in the App Store because of the CLA.

**Comment by Semisol** [comment]
> which sadly makes Audacity the only viable option in the App Store because of the CLA

I don't think it is a big deal. I don't find it too hard to search it up, download an image and drag something.

**Comment by TheEvilSkeleton** [comment]
As apps become more accessible and much easier to search in a centralized store like the App Store, I think it'll become worse and worse for us as time goes by. We can search through categories inside the app store which makes it many times easier to find apps, unlike searching on the web. But I digress and I don't want to derail the current discussion.

---

Since we have very little knowledge with XDG portals, we could ask in [`#flatpak:matrix.org` ](https://matrix.to/#/#flatpak:matrix.org) about how to implement it.

**Comment by TheEvilSkeleton** [comment]
Reading through https://github.com/tenacityteam/tenacity/issues/438#issuecomment-899115260, I guess we should make it optional.

**Comment by n0toose** [comment]
> Based on https://developer.apple.com/forums/thread/18922, we cannot release it in the App Store, which sadly makes Audacity the only viable option in the App Store because of the CLA.

I completely forgot about how Apple has hated GPL software since 2009, sorry.

**Comment by leio** [comment]
> @leio I forgot to mention that we'll probably have to use libportal temporarily because, based on my friends, they said:
> 
> > yeah looks like wxWidgets just uses GtkFileChooserDialog not GtkFileChooserNative
> > so you'd have to use libportal
> 
> If we manage to send the relevant patches to upstream, we might have FileChooser portal without needing to use libportal, from what I understood.

I intend to look at adding GtkFileChooserNative support to wxWidgets upstream and update the local copy in src/widgets/FileDialog/gtk/ with such a version, until we can depend on a wxWidgets stable version that includes these changes.
